File: python-backend/app/schemas/task.py
from typing import Annotated, Literal

from pydantic import BaseModel, StringConstraints, model_validator

TaskPriority = Literal["low", "medium", "high"]
TrimmedTitle = Annotated[str, StringConstraints(strip_whitespace=True, min_length=1, max_length=200)]
TrimmedDueDate = Annotated[str, StringConstraints(strip_whitespace=True, max_length=30)]
TrimmedNotes = Annotated[str, StringConstraints(strip_whitespace=True, max_length=5000)]


class CreateTaskRequest(BaseModel):
    title: TrimmedTitle
    completed: bool | None = None
    priority: TaskPriority = "medium"
    dueDate: TrimmedDueDate | None = None
    notes: TrimmedNotes | None = None


class UpdateTaskRequest(BaseModel):
    title: TrimmedTitle | None = None
    priority: TaskPriority | None = None
    dueDate: TrimmedDueDate | None = None
    notes: TrimmedNotes | None = None
    completed: bool | None = None
    version: int

    @model_validator(mode="after")
    def ensure_mutation_fields(self) -> "UpdateTaskRequest":
        changed_fields = {name for name in self.model_fields_set if name != "version"}
        if not changed_fields:
            raise ValueError("At least one field must be provided")
        return self


class TaskResponse(BaseModel):
    id: str
    title: str
    completed: bool
    priority: TaskPriority
    dueDate: str | None = None
    notes: str | None = None
    createdAt: int
    updatedAt: int
    version: int


class TaskEnvelope(BaseModel):
    data: TaskResponse
    source: str


class TaskListEnvelope(BaseModel):
    items: list[TaskResponse]
    total: int
    source: str

File: src/core/types/task.ts
/**
 * 任务相关类型定义
 */

import { BaseEntity } from './base';

/** 任务优先级 */
export type TaskPriority = 'low' | 'medium' | 'high';

/**
 * 任务实体
 */
export interface Task extends BaseEntity {
  title: string;
  completed: boolean;
  priority: TaskPriority;
  dueDate?: string; // ISO 日期格式 YYYY-MM-DD
}

/**
 * 创建任务 DTO
 */
export interface CreateTaskDTO {
  title: string;
  completed?: boolean;
  priority?: TaskPriority;
  dueDate?: string;
}

/**
 * 更新任务 DTO
 */
export type UpdateTaskDTO = Partial<CreateTaskDTO>;

/**
 * 任务统计结果
 */
export interface TaskStats {
  total: number;
  completed: number;
  pending: number;
}
